### ISSUE TITLE
Ember 1.13 How to add classes to the root view?

### ISSUE THREAD

**Issue by OrKoN** [emberjs/ember.js#11486]
The following was the way to style the root ember-view of an app http://emberjs.jsbin.com/yasukevoni/1/edit?html,js,console,output  in Ember 1.13

Now Views are deprecated and I tried to follow the deprecation guide and migrate from using ApplicationView to using a component http://emberjs.jsbin.com/hekadobewa/edit?html,js,console,output But the new way is not equivalent to the old one because an extra view element is added:

Expected structure (from the first jsbin):
![image](https://cloud.githubusercontent.com/assets/399150/8189273/76d425de-145b-11e5-9a37-3ccecbaa98e5.png)

Actual structure (from the second jsbin):
![image](https://cloud.githubusercontent.com/assets/399150/8189428/82db68e6-145c-11e5-8e29-70d088957f04.png)

How to migrate this case to Ember 1.13?

**Comment by joeruello** [comment]
I think this functionality has been lost. I assume we'll be able to create an `application` component when [Routable Components](https://github.com/ef4/rfcs/blob/routeable-components/active/0000-routeable-components.md) are implemented. 

A workaround until this is implemented may be to put the element with desired classes in your index.html, and then [change the root element](http://guides.emberjs.com/v1.10.0/configuring-ember/embedding-applications/).

**Comment by sarus** [comment]
I'm confused by this too.  Previously (< Ember 1.12.0) I was using views to add a class to the view's root DOM element.  Something like this:

```
import Ember from "ember";

export default Ember.View.extend({
  classNames: ['account-view']
});
```

Now that views are deprecated I removed all of these view classes and just wrapped the "view" in the handlebars template with a div and added the class.  

```
<div class="account-view">
{{!-- Existing Template Code --}}
</div>
```

After removing all views and getting rid of the deprecations (I even deleted the views folder out of my project) I noticed that the top level of the app is still wrapped by a `div` with the class `ember-view`.  

Should this be gone now that views are deprecated or at least this class should be renamed? 

 Is it safe to apply a style to ember-view under the assumption that there will only be one in the application assuming I got rid of all my own views?

Thanks!

**Comment by sarus** [comment]
Just answering my own question you cannot style the class `ember-view` under the assumption that it only exists in the top level DIV wrapping the application as it shows up in components as well.

**Comment by Samsinite** [comment]
Couldn't one just be very specific in their selectors for styling:

```css
body > .ember-view {
  ...
}
```

Or does this not work for styling the root ember view in the application?

**Comment by sarus** [comment]
@Samsinite that does work =)  I should have thought of it.  So simple to fix haha.  I would still argue that the class `ember-view` should be renamed since views are no longer supposed to exist though.

**Comment by stefanpenner** [comment]
@rwjblue let's add this today's agenda

**Comment by topaxi** [comment]
I just started on an application this Monday and needed to add an application view to add/change classes too :+1:

**Comment by mixonic** [comment]
In 2.0 there will not be a solution for the specific case discussed here. There will be no way to add classes to the application view's element. You **MUST** use an additional DOM node inside the application template.

We discussed the issue in some detail today, and it seems this is a use case that will remain awkward until we land routable components. For users who absolutely require customization of the application view's element, I suggest using the [ember-legacy-views](https://github.com/emberjs/ember-legacy-views) addon. In 1.13, that addon will silence deprecation warnings regarding views, in 2.0 it will cause the API to continue being present. After routable components land, you should be able to safely migrate away from that addon.

I'm going to close this issue for now as a wontfix, but please feel free to comment or re-open if I have missed something significant!

**Comment by OrKoN** [comment]
@mixonic It looks like it's indeed possible in 2.0 http://stackoverflow.com/a/32013110/601499 Could you please comment if it can serve as a solution or if it should be avoided?

**Comment by rwjblue** [comment]
Using a component as ApplicationView will allow customization of classNames and whatnot, but is definitely going to have a number of negative results as well (for example controller is not correct).

This is definitely not a go forward solution, and there is nothing wrong with using the addon if you require this functionality.

**Comment by OrKoN** [comment]
@rwjblue Thanks for the comment! But what do you mean by controller is incorrect? I tried to access controller's properties and it looks ok to me: http://emberjs.jsbin.com/napedapuju/1/edit?html,js,console,output Anyway I have updated my answer with your comment.

**Comment by DanChadwick** [comment]
A few gotchas. First, there may be other top-level div's of class ember-view. For example, both ember-wormhole and ember-paper's paper-menu place divs at the top level (although only the later has this class). This makes selecting only the application view hierarchy with CSS difficult.

Using rootElement had negative side effects on ember-paper's menu handling, since it moves the listener down from the body to whatever div you create.

This leaves the legacy views add-on or difficult CSS classes until you can customize the root component in whatever version has routable components. Ugh.

**Comment by davidlormor** [comment]
Just wanted to second the comments from @DanChadwick (and everyone else) - I'm using flexbox for much of my app layout, and not being able to target the root `div` of the app via a simple class or ID is a pain.

**Comment by Gaurav0** [comment]
@rwjblue @stefanpenner @mixonic Is it possible, given the loss of this functionality, to continue to support ember-legacy-views until Routable Components or some other way of implementing this lands?

**Comment by mixonic** [comment]
I need to state first that this feature (classes on the root) is *not a feature of Ember 2.x*. It was removed after 1.13, and though we support it via the legacy addon it is not in 2.x proper. But technicalities be damned, I agree the practicalities are most important here :-)

@Gaurav0 it is difficult to say that every stable release of Ember will have this feature between the loss of support for ember-legacy-views (when the glimmer refactor lands) and when we add support for routable components. I'll make sure people on core have it in mind (it impacts their apps as well) and see if there is any specific thought. I think it is safe to suggest we all want it to be possible to add classes on the root in each release of the LTS channel.

**Comment by himedlooff** [comment]
It would be really nice to not have to use `body > .ember-view`.

**Comment by mellatone** [comment]
This is extremely important.

**Comment by heyjinkim** [comment]
I'm having the issue with styling `body > .ember-view` when using `ember-modal-dialog` addon because the modal-dialog element is appended to the body tag that breaks the `body > .ember-view` style. This is hacky, but here's how I fixed.

app/index.html
```
<body>
  <div id="ember-application"></div>
</body>
```

config/environment.js
```
ENV.APP.rootElement = '#ember-application';
```

app.css
```
#ember-application > .ember-view {
  // do your style
}
```

**Comment by erichonkanen** [comment]
@heyjinkim I think your answer is effectively the same as doing `body > .ember-view {}`

**Comment by topaxi** [comment]
@erichonkanen This may conflict with some addons, which append components into the `body`.

I ran into this issue myself and I am using the following selector: `body > .ember-view:first-of-type`

**Comment by Dynasty9** [comment]
I've got the pod structure but it should be the same for you guys, Under pods/application/view.js, I have Ember.Component.extend instead of Ember.View.extend. It looks like this:
```
import Ember from 'ember';
export default Ember.Component.extend({
    classNames : ["fill-space","viewport-table"]
});
```

**Comment by amk221** [comment]
@Dynasty9 Thanks for the solution.
Any problems with adding `tagName: ''`?

**Comment by rwjblue** [comment]
FWIW, the suggestion from https://github.com/emberjs/ember.js/issues/11486#issuecomment-225255269 will not work in Ember 2.7.

**Comment by amk221** [comment]
revert revert :) Thank you for the heads up.

**Comment by Dynasty9** [comment]
👍 noted

**Comment by kanongil** [comment]
Maybe reopen this issue now that ember-legacy-views is gone and routeable components are still MIA?

Currently I am doing it in `application/view.js` like this:
```js
const TopLevelView = MyApp.__container__.lookupFactory('view:toplevel');

export default TopLevelView.extend({
  classNames: 'my-app-main'
});
```

**Comment by annamalai-saro** [comment]
When using `flexbox` , the layout breaks because of parent-child relationship

**Comment by erichonkanen** [comment]
@kanongil I just created an `{{app-main}}` component and it does the trick...

**Comment by justinaray** [comment]
@rwjblue you mentioned above that the trick of exporting a Component from view.js would not work in 2.7.  However, it still seems to be fine in our 2.7 upgrade branch.  Was this something that you wanted to address but didn't land in 2.7?  Or is there a subtlety that isn't working that we're not noticing?

We don't like using this trick, but it has deferred the app-level styling rework we'd need to take on to account for the extra div in our dom.  In this case, we don't need any controller guarantees ... just a hook for the styles.

**Comment by amk221** [comment]
Upgraded to 2.8.0 now, and obviously the `tagName: ''` view fix no longer works.
Wondering will this will be officially possible soon?
It's frustrating having to do this:
```
:root,
.ember-application,         // Class name Ember add's the the body
.ember-application > div {  // Extraneous div
  height: 100%;
}
```

**Comment by rohmann** [comment]
Ran into this today after updating. After fiddling with it, I got it to accept a component via using an instance initializer.

https://gist.github.com/rohmann/1df36a63f2c3905bc52ae373e771a129

**Comment by Redsandro** [comment]
@kanongil 
> Currently I am doing it in `application/view.js` like this:

Can you elaborate? Where is this `application` directory? What is `MyApp`?

**Comment by kanongil** [comment]
@Redsandro This method doesn't work any more.

**Comment by Kilowhisky** [comment]
Still not a workable solution? Its really quite difficult to do the height: 100% trick.

html, body {
     height: 100%;
}

etc...

**Comment by humanchimp** [comment]
@Samsinite I realize you wrote that comment a very long time ago, but that solution only works if your root element is the only `.ember-view` in the body (which in our case, it is not)

**Comment by ChrizzDF** [comment]
Hey guys,
I ran into the same problem with some addon components why
`body > .ember-view {}` `body > .ember-view:first-of-type` didn't work.

I found a solution (at least for me):
'body > [class="ember-view"] {}`
This affects only the desired element because other addons always have additional classes besides `.ember-view`

Hopefully someone finds this also useful!

**Comment by blufish** [comment]
Amazing that Ember devs would not address this. Very ridiculous that you cannot control the controllers, must use them, and have Ember create a "ember-view" in the DOM. 

Why do controllers not follow the component spec?

**Comment by kanongil** [comment]
This issue should be resolved once https://github.com/emberjs/rfcs/pull/280 lands.

**Comment by blufish** [comment]
Excellent!! Sorry I didn't realize that was coming!
